rust/lance-index/src/vector/ivf.rs

```diff
@@ -17,39 +17,211 @@
 use std::ops::Range;
 use std::sync::Arc;
 
+use arrow_array::types::{Float16Type, Float32Type, Float64Type};
 use arrow_array::{
-    cast::AsArray,
-    types::{Float32Type, UInt32Type},
-    Array, Float32Array, RecordBatch, UInt32Array,
+    cast::AsArray, types::UInt32Type, Array, FixedSizeListArray, RecordBatch, UInt32Array,
 };
 use arrow_ord::sort::sort_to_indices;
-use arrow_schema::Field;
+use arrow_schema::{DataType, Field};
 use arrow_select::take::take;
-use lance_arrow::RecordBatchExt;
+use async_trait::async_trait;
+use lance_arrow::*;
 use lance_core::{Error, Result};
 use lance_linalg::{
-    distance::{cosine_distance_batch, dot_distance_batch, l2_distance_batch, MetricType},
+    distance::{
+        cosine_distance_batch, dot_distance_batch, l2_distance_batch, Cosine, Dot, MetricType, L2,
+    },
     MatrixView,
 };
 use snafu::{location, Location};
 use tracing::instrument;
 
+mod builder;
+
 use super::{PART_ID_COLUMN, PQ_CODE_COLUMN, RESIDUAL_COLUMN};
 use crate::vector::{
     pq::{transform::PQTransformer, ProductQuantizer},
     residual::ResidualTransform,
     transform::Transformer,
 };
+pub use builder::IvfBuildParams;
+
+fn new_ivf_impl<T: ArrowFloatType + Dot + Cosine + L2 + 'static>(
```

Comment:
If we find ourselves repeating `ArrowFloatType + Dot + Cosine + L2 + 'static` a lot we might want to make an alias:

```
pub type LanceFloatType = ArrowFloatType + Dot + Cosine + L2 + 'static;
```

Reply:
ok lemme try it

Reply:
Could not make it work, `ArrowFloatType` is a template here. @wjones127 any suggestion?

Reply:
Ah, this works...but maybe this just adds more confusion:

```
pub trait LanceFloatType: ArrowFloatType + Dot + Cosine + L2 + 'static {}
impl<T: ArrowFloatType + Dot + Cosine + L2 + 'static> LanceFloatType for T {}
```

Reply:
Also https://doc.rust-lang.org/beta/unstable-book/language-features/trait-alias.html but given its been six years I wouldn't hold my breath :laughing: